### ISSUE TITLE
Open link in new background tab (F / alt-f) actually opens tab in foreground (Chrome bug)!

### ISSUE THREAD

**Issue by jcvfen** [philc/vimium#1279]
doesnt open in background anymore.

Chrome dev
Version 41.0.2224.3 dev-m (64-bit)

**Comment by TheQwerty** [comment]
There's a few issues on the Chromium tracker that might cover this but right now this one seems to have the most promise: https://code.google.com/p/chromium/issues/detail?id=454222

As an extension, Vimium could make use of the [chrome.tabs.create](https://developer.chrome.com/extensions/tabs#method-create) API instead of simulating clicks, but then any onclick code from active site wouldn't get triggered. :confused:

**Comment by mrmr1993** [comment]
[454222](https://code.google.com/p/chromium/issues/detail?id=454222) is exactly this issue, thanks @TheQwerty. Anyone experiencing this problem should go there and star it, so the team know it's affecting people and focus developer effort there.

> ... but then any onclick code from active site wouldn't get triggered.

This would remove a lot (I'd say most) of the value from the extension - we'd become instantly unable to control any page that uses javascript (Facebook, Google+, GMail, parts of Reddit and GitHub, etc.). It's probably not worth crippling the extension to fix a minor annoyance on the Chrome dev branch.

**Comment by mrmr1993** [comment]
It looks like **jochen@** is taking a stance *against* the old behaviour over in [454222](https://code.google.com/p/chromium/issues/detail?id=454222), so we might have to take action to stop this breaking for all users. Since it's already hit the dev branch and the release date for Chrome 41 is approaching, I think we should be reasonably concerned.

The only fix that I can think of is:
* modify `DomUtils.simulateClick` to
  - add a conditional `mouseEvent.preventDefault()` call before `element.dispatchEvent(mouseEvent)`
  - return whether `preventDefault` was called on the event:

```coffee
  simulateClick: (element, modifiers, preventDefault = false) ->
    ...
      mouseEvent = document.createEvent "MouseEvents"
      ...
      if preventDefault
        defaultPrevented = false # Tracks whether any handlers call preventDefault.
        mouseEvent.preventDefault()
        mouseEvent.preventDefault = ->
          defaultPrevented = true
          undefined
      ...
      element.dispatchEvent mouseEvent
      # Return whether any listeners called preventDefault if we'd already called it.
      if preventDefault then defaultPrevented else undefined
```
* Update `openUrlInNewTab` in background_scripts/main.coffee so that it respects `request.selected` (when it's specified)
  - chose `selected` because it's already used in `toggleViewSource` in vimium_frontend.coffee.
* Update `LinkHints.setOpenLinkMode` to prevent the default action, and use the extension API to open a link in a new tab. This can be done by changing the `@linkActivator` defined for `OPEN_IN_NEW_BG_TAB`, `OPEN_IN_NEW_FG_TAB` and `OPEN_WITH_QUEUE` to:

```coffee
      @linkActivator = (link) ->
        # When "clicking" on a link, dispatch the event with the appropriate meta key (CMD on Mac, CTRL on
        # windows) to open it in a new tab if necessary.
        #
        # NOTE(mrmr1993): It is no longer possible to open links in the background using a simulated click.
        # See https://code.google.com/p/chromium/issues/detail?id=454222 for details.
        #
        # To work around this, we simulated a ctrl-click, prevent the default action, and open a
        # backgrounded tab with the URL using chrome.tabs.create.
        preventDefault = @mode isnt OPEN_IN_NEW_FG_TAB and link.tagName is "A" and
                         link.href.substr(0,11) isnt "javascript:"
        defaultPrevented = DomUtils.simulateClick link, {
          shiftKey: @mode is OPEN_IN_NEW_FG_TAB,
          metaKey: KeyboardUtils.platform == "Mac",
          ctrlKey: KeyboardUtils.platform != "Mac",
          altKey: false}, preventDefault
        if preventDefault and not defaultPrevented
          chrome.runtime.sendMessage { handler: "openUrlInNewTab", url: link.href, selected: false }
```

I haven't tested this (and obviously I hope we're not going to need it), but I think it should emulate the current behaviour on Chrome 40 ([jsfiddle for testing](http://jsfiddle.net/70zmsqsh/1/)). Any thoughts guys?

**Comment by mrmr1993** [comment]
@smblott-github can you update the title for this issue? At the moment it has quite low discoverability, so we seem to be getting duplicate issues.

Would recommend something like "Open link in new background tab (F / alt-f) focuses the new tab".

**Comment by gavinwahl** [comment]
I am experiencing this as well.

**Comment by altern8tif** [comment]
Same issue on the latest stable release of Chrome 41.0.2272.76.

**Comment by vibhav** [comment]
FWIW, I'm running into bug this as well after upgrading to Version 41.0.2272.76. Let us know if there's anything we can do to help.

**Comment by mohitleo9** [comment]
Yeah, it's strange it started happening some time ago.

**Comment by smblott-github** [comment]
@gavinwahl, @altern8tif, @vibhav, @mohitleo9. It looks like this is caused by this [Chrome bug](https://code.google.com/p/chromium/issues/detail?id=454222).  The more people that go over there and "star" the bug the better.  It's not clear yet whether Chromium folks will actually address the issue.

**Comment by vibhav** [comment]
@smblott-github Thanks for the pointer. Done.

**Comment by charle5** [comment]
ditto, and starred bug `Version 41.0.2272.76 (64-bit)`

**Comment by dwayneyuen** [comment]
Starred the bug linked. Thanks!
Chrome: Version 41.0.2272.76 (64-bit)
Vimium: Version 1.49

**Comment by smblott-github** [comment]
Looks like this has been [fixed](https://code.google.com/p/chromium/issues/detail?can=2&start=0&num=100&q=&colspec=ID%20Pri%20M%20Week%20ReleaseBlock%20Cr%20Status%20Owner%20Summary%20OS%20Modified&groupby=&sort=&id=454222) in Chromium, and will start filtering through soon.

**Comment by fevrin** [comment]
Odd, I wasn't affected by this bug until upgrading today to Chromium 41.0.2272.76-0ubuntu0.14.10.1.1118.  Hopefully the new build does come down soon :3

**Comment by Gpx** [comment]
Does anyone know the next Chrome release date?

**Comment by altern8tif** [comment]
@Gpx Seems like it has been [fixed] (https://code.google.com/p/chromium/issues/detail?id=454222#c36) in Chromium and merged into the 2311 release (currently released in beta 2 days ago).

Unfortunately, I'm still experiencing the bug in Chrome Beta 42.0.2311.39.  Anyone else has the same experience?

**Comment by smblott-github** [comment]
@altern8tif. The merge was only approved 7 hours ago (so at about 8:00 UTC, 2015-03-13). It'll take some time to filter through.

**Comment by emallove** [comment]
The following Chrome extension provides a workaround: [Force Background Tab](https://chrome.google.com/webstore/detail/force-background-tab/gidlfommnbibbmegmgajdbikelkdcmcl/related).

**Comment by mrmr1993** [comment]
@emallove it looks like that extension forces *every* tab (except `chrome:` and some other internal type urls) to open in a background tab. This isn't a true workaround, but rather breaks the foreground opening shortcut rather than the background one.

I haven't been able to test this, but, from [the source](https://github.com/monque/ChromeExtension-ForceBackgroundTab/blob/master/extension/script/background.js) it also looks like it might background new tabs which aren't the expected `chrome://newtab`, and so may cause unpleasantness with eg. newtab-overriding extensions or our Vimium's new tab page override functionality.

This should be fixed anyway in the next chrome release (before the next version, but with no certain release date as yet). This should be within the next few days.

**Comment by TheQwerty** [comment]
FYI, all is well using Chrome version 43.0.2333.3 canary (64-bit).

I imagine it's also fixed in the next dev and beta releases since they made the fix part of the M-42 milestone, but still waiting on those to drop.

Users on the stable branch probably won't see this fixed until v42 is promoted, which I believe won't be until early April.

**Comment by arizzubair** [comment]
When can we expect updated version of vimium without this bug? I am currently using chrome on windows (build: 41.0.2272.118 m)

**Comment by mrmr1993** [comment]
This is a Chrome bug, which the Chrome 42 release will fix. It will be released later this month, but you can switch to the dev channel to take advantage of the fix now.

**Comment by gavinwahl** [comment]
Running 42.0.2311.68, this is not completely fixed. Normal links open in a background tab, but links with `target="_blank"` open in the foreground (despite opening in the background when middle-clicked). 

I've posted a test case here: http://jsbin.com/raxorevane/1/. The behavior of hinting a link with `F` should be the same as middle-clicking it.

**Comment by mrmr1993** [comment]
Presumably ctrl-clicking also opens it in a background tab? Opening a Chromium issue for this is way to go.

**Comment by mohitleo9** [comment]
Finally it's fixed, my chrome just updated and I was really surprised to see links opening in new tabs (without the current page loosing focus).

**Comment by wocala** [comment]
Some websites are fixed, but this one is not. 
http://www.zhihu.com/
It is a Chinese website.

**Comment by mrmr1993** [comment]
@wocala Does ctrl-clicking the links on that page open them in a new tab?

**Comment by altern8tif** [comment]
@mrmr1993 @wocala 

It seems that this is an issue only when the links have the attribute ``target="_blank"``.  For such links, Ctrl/Cmd-clicking it opens it in a new background tab, but Vimium opens it in a new foreground tab.  Also mentioned by @gavinwahl https://github.com/philc/vimium/issues/1279#issuecomment-90252679

**Comment by mrmr1993** [comment]
We don't seem to be treating them differently anywhere in the code, so presumably this is a Chromium bug. Can you follow the instructions in [this comment] (https://github.com/philc/vimium/issues/1279#issuecomment-64960158) to confirm this?

**Comment by altern8tif** [comment]
@mrmr1993 Just ran the code in the devtools console.  Yup, the links with ``target="_blank"`` open up in a new foreground tab.  Is this the same Chromium bug as before?

**Comment by mrmr1993** [comment]
> Is this the same Chromium bug as before?

Maybe, maybe not; depends on what's causing the difference. It's probably best to open a new one and refer to the last one in it.

**Comment by gdh1995** [comment]
`target="_blank"` still makes a new tab foreground even we set `ctrlKey: true`.

Here's an example: http://output.jsbin.com/raxorevane/1/edit?html,output , though I find vimium do not work on the output iframe, since it has no attr `url`.

**Comment by mrmr1993** [comment]
I've opened [Chromium issue 486711](https://code.google.com/p/chromium/issues/detail?id=486711) for this specific case.

**Comment by zhangkaifei** [comment]
mac os 10.10.2    chrome  42.0.2311.135 (64-bit)    Ctrl/Cmd-clicking it opens it in a new background tab, but Vimium opens it in a new foreground tab

**Comment by wocala** [comment]
It's 43.0.2357.124 (stable) m （32 bit）now. Problem still remains. Try this website:
https://www.google.com/?gws_rd=ssl#newwindow=1&q=galvanic+replacement+reaction
Is there any new update about this problem?

**Comment by gdh1995** [comment]
I think we may force `<a>`: `.attr("target", "_blank")`, and reset `target` to its previous value.

**Comment by smblott-github** [comment]
@wocala.  I don't quite follow you here.  I go to that site... Do what? What happens? What do think should happen?  Thanks.

**Comment by mrmr1993** [comment]
@wocala the fix was merged into Chrome 44, which should land land any day now, according to the Chromium schedule. (Because of this, we should probably leave the code alone and just wait for the Chromium-side fix to land @gdh1995).

@smblott-github [This comment](https://github.com/philc/vimium/issues/1279#issuecomment-98478118) describes the issue best here. [Chromium issue 486711](https://code.google.com/p/chromium/issues/detail?id=486711) is the corresponding Chromium issue.

**Comment by wocala** [comment]
@mrmr1993 Thanks for your help!

**Comment by wocala** [comment]
@smblott-github I thought it's quiet clear here.
As the title of this issue shows: "Open link in new background tab (F / alt-f) actually opens tab in foreground", when I want to open new background tab in the web I pasted, it opens in foreground. 
I saw the last comment of this issue is quite long ago and problem remains, so I want to know the process of this problem. That's what I mean. Thanks.

**Comment by rajeevksahu** [comment]
Being the efficient packers & movers in Bangalore , we are always at your support for providing extremely qualitative moving alternatives for your urgent need. Our packers movers Mangalore alternatives are much in requirement as we provide our alternatives within the shortest possible time which is much appreciated by a variety of people. 

<a href="http://www.expert5th.in/packers-and-movers-hyderabad/" title="Movers and Packers in Hyderabad">Movers and Packers in Hyderabad</a>
<a href="http://www.expert5th.in/packers-and-movers-bangalore/" title="Movers and Packers in Bangalore">Movers and Packers in Bangalore</a>
<a href="http://www.expert5th.in/packers-and-movers-mumbai/" title="Movers and Packers in Mumbai">Movers and Packers in Mumbai</a>

**Comment by rajeevksahu** [comment]
Besides this we have a huge storage service which is much useful for storing your useful items to be securely delivered to your place. We proffer a lot of packaging and movers to our customers, which are just designed keeping in view of their moving specifications and needs. At our packaging and movers, we are all set to provide you a variety of moving alternatives to you such as house moving, domestic moving, corporate moving, company moving, workplace moving and so on. 

<a href="http://www.expert5th.in/packers-and-movers-gurgaon/" title="Movers and Packers in Gurgaon">Movers and Packers in Gurgaon</a>
<a href="http://www.expert5th.in/packers-and-movers-delhi/" title="Movers and Packers in Delhi">Movers and Packers in Delhi</a>
<a href="http://www.expert5th.in/packers-and-movers-pune/" title="Movers and Packers in Pune">Movers and Packers in Pune</a>

**Comment by rajeevksahu** [comment]
Each and every moving support is provided to you cheaply which is the reason why every individual wants to acquire our alternatives. We always deal with the moving specifications of our customers while giving them our support. Being a well-known packaging and moving organizations, our company has handled great standards in the field of packaging and movers. 

<a href="http://www.expert5th.in/packers-and-movers-chennai/" title="Movers and Packers in Chennai">Movers and Packers in Chennai</a>
<a href="http://www.expert5th.in/packers-and-movers-noida/" title="Movers and Packers in Noida">Movers and Packers in Noida</a>
<a href="http://www.expert5th.in/packers-and-movers-navimumbai/" title="Movers and Packers in Navi Mumbai">Movers and Packers in Navi Mumbai</a>
<a href="http://www.expert5th.in/packers-and-movers-thane/" title="Movers and Packers in Thane">Movers and Packers in Thane</a>
<a href="http://www.expert5th.in/packers-and-movers-ghaziabad/" title="Movers and Packers in Ghaziabad">Movers and Packers in Ghaziabad</a>
<a href="http://www.expert5th.in/packers-and-movers-faridabad/" title="Movers and Packers in Faridabad">Movers and Packers in Faridabad</a>

**Comment by rajeevksahu** [comment]
We never are unsuccessful of the best features to provide qualitative packaging and movers to our clients, which is why they like our alternatives. All our moving alternatives are carried out in our company devising the existing advanced methods. With the help of these methods we are able to provide your valuables in a secure way. 

<a href="http://www.expert5th.in/packers-and-movers-bangalore/" title="Packers and Movers in Bangalore">Packers and Movers in Bangalore</a>
<a href="http://www.expert5th.in/packers-and-movers-pune/" title="Packers and Movers in Pune">Packers and Movers in Pune</a>
<a href="http://www.expert5th.in/packers-and-movers-mumbai/" title="Packers and Movers in Mumbai">Packers and Movers in Mumbai</a>
